contracts/src/RIP7755Source.sol

```diff
@@ -0,0 +1,271 @@
+    /// @param status The `CrossChainCallStatus` status that the request has
+    error InvalidStatusForRequestCancel(CrossChainCallStatus status);
+
+    /// @notice This error is thrown if a Filler attempts to claim a reward for a request that is not in a `CrossChainCallStatus.Requested` state
+    /// @param status The `CrossChainCallStatus` status of the request
+    error InvalidStatusForClaim(CrossChainCallStatus status);
+
+    /// @notice This error is thrown if an attempt to cancel a request is made before the request's expiry timestamp
+    /// @param currentTimestamp The current block timestamp
+    /// @param expiry The timestamp at which the request expires
+    error CannotCancelRequestBeforeExpiry(uint256 currentTimestamp, uint256 expiry);
+
+    /// @notice Submits an RIP-7755 request for a cross chain call
+    ///
+    /// @param request A cross chain request structured as a `CrossChainRequest`
+    function requestCrossChainCall(CrossChainRequest memory request) external payable {
+        request.nonce = ++_nonce;
+        bool usingNativeCurrency = request.rewardAsset == _NATIVE_ASSET;
+        uint256 expectedValue = usingNativeCurrency ? request.rewardAmount : 0;
+
+        if (msg.value != expectedValue) {
+            revert InvalidValue(expectedValue, msg.value);
+        }
+
+        bytes32 requestHash = hashRequestMemory(request);
+        _requestMetadata[requestHash] = RequestMeta({
+            status: CrossChainCallStatus.Requested,
+            expiryTimestamp: uint40(block.timestamp + request.validDuration),
+            requester: msg.sender
+        });
+
+        emit CrossChainCallRequested(requestHash, request);
+
+        if (!usingNativeCurrency) {
+            _pullERC20({owner: msg.sender, asset: request.rewardAsset, amount: request.rewardAmount});
+        }
+    }
+
+    /// @notice To be called by a Filler that successfully submitted a cross chain request to the destination chain and
+    /// can prove it with a valid nested storage proof
+    ///
+    /// @param request A cross chain request structured as a `CrossChainRequest`
+    /// @param fillInfo The fill info that should be in storage in `RIP7755Verifier` on destination chain
+    /// @param storageProofData A storage proof that cryptographically verifies that `fillInfo` does, indeed, exist in
+    /// storage on the destination chain
+    /// @param payTo The address the Filler wants to receive the reward
+    function claimReward(
+        CrossChainRequest calldata request,
+        RIP7755Verifier.FulfillmentInfo calldata fillInfo,
+        bytes calldata storageProofData,
+        address payTo
+    ) external {
+        bytes32 requestHash = hashRequest(request);
+        bytes32 storageKey = keccak256(
+            abi.encodePacked(
+                requestHash,
+                uint256(0) // Must be at slot 0
```

Comment:
Please go through PR comments and incorporate useful feedback, i.e. https://github.com/ethereum/RIPs/pull/31#discussion_r1721909192